ext/standard/string.c

```diff
@@ -2184,14 +2184,12 @@ PHP_FUNCTION(substr)
 {
 	zend_string *str;
 	zend_long l = 0, f;
+	zend_bool len_is_null = 1;
 	int argc = ZEND_NUM_ARGS();
 
-	ZEND_PARSE_PARAMETERS_START(2, 3)
-		Z_PARAM_STR(str)
-		Z_PARAM_LONG(f)
-		Z_PARAM_OPTIONAL
-		Z_PARAM_LONG(l)
```

Comment:
You can use `Z_PARAM_LONG_EX(l, len_is_null, 1, 0)` here to avoid switching API.

(We should probably add some convenience macros like Z_PARAM_LONG_OR_NULL.)

Reply:
Looks good. I'd suggest to also replace the existing uses of Z_PARAM_LONG_EX with the new macro.

Reply:
Done!